runners/core-java/src/test/java/org/apache/beam/runners/core/ReduceFnRunnerTest.java

```diff
@@ -873,6 +907,288 @@ public void testWatermarkHoldAndLateData() throws Exception {
     tester.assertHasOnlyGlobalAndFinishedSetsFor();
   }
 
+  @Test
+  public void testMergingWatermarkHoldAndLateDataSpecific() throws Exception {
```

Comment:
This test should be simpler and shorter.

Reply:
Would you prefer:
- a test helper function taking configuration objects with separate tests for each configuration
- remove most of these and just keep a complicated one
- making a lot of separate tests but removing configuration object and just duplicating test setup

These were written at the same time as I detected this issue but are unrelated. They seem like useful additional coverage but I could also put them into a separate change if you'd prefer.

Reply:
I like the first option better. In case a test fails, we will know which exact config failed and it will be easier to fix. I saw a post (TotT) about having loops in tests is generally a bad idea.

Reply:
Done